contrib/opentimelineio_contrib/adapters/otio_svg.py

```diff
@@ -0,0 +1,481 @@
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the Apache License with the above modification is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied. See the Apache License for the specific
+# language governing permissions and limitations under the Apache License.
+#
+
+"""OTIO to SVG Adapter"""
+
+import opentimelineio as otio
+
+from random import seed
+from random import random
+
+seed(50)
+
+
+class Color:
+    r = 0.0
+    g = 0.0
+    b = 0.0
+    a = 0.0
+
+    def __init__(self, r=0.0, g=0.0, b=0.0, a=0.0):
+        self.r = r
+        self.g = g
+        self.b = b
+        self.a = a
+
+    @staticmethod
+    def random_color():
+        return Color(random(), random(), random(), 1.0)
+
+
+transparent = Color(0, 0, 0, 0)
+black = Color(0.0, 0.0, 0.0, 1.0)
+white = Color(1.0, 1.0, 1.0, 1.0)
+transluscent_white = Color(1.0, 1.0, 1.0, 0.7)
+purple = Color(0.5, 0.0, 0.5, 1.0)
+light_blue = Color(0.529, 0.808, 0.922, 1.0)
+blue = Color(0.0, 0.0, 1.0, 1.0)
+dark_blue = Color(0.0, 0.0, 0.54, 1.0)
+green = Color(0.0, 0.5, 0.0, 1.0)
+dark_green = Color(0.0, 0.39, 0.0, 1.0)
+yellow = Color(1.0, 1.0, 0.0, 1.0)
+gold = Color(1.0, 0.84, 0.0, 1.0)
+orange = Color(1.0, 0.647, 0.0, 1.0)
+red = Color(1.0, 0.0, 0.0, 1.0)
+dark_red = Color(0.54, 0.0, 0.0, 1.0)
+brown = Color(0.54, 0.27, 0.1, 1.0)
+pink = Color(1.0, 0.75, 0.79, 1.0)
+gray = Color(0.5, 0.5, 0.5, 1.0)
+dark_gray = Color(0.66, 0.66, 0.66, 1.0)
+
+
+class Point:
+    x = 0.0
+    y = 0.0
+
+    def __init__(self, x=0.0, y=0.0):
+        self.x = x
+        self.y = y
+
+
+class Rect:
+    origin = Point()
+    width = 0.0
+    height = 0.0
+
+    def __init__(self, origin=Point(), width=0.0, height=0.0):
+        self.origin = origin
+        self.width = width
+        self.height = height
+
+    def normalized(self):
+        normalized_origin = Point(
+            self.origin.x + (self.width if self.width < 0 else 0),
+            self.origin.y + (self.height if self.height < 0 else 0),
+        )
+        normalized_width = abs(self.width)
+        normalized_height = abs(self.height)
+        return Rect(normalized_origin, normalized_width, normalized_height)
+
+    def min_x(self):
+        return self.normalized().origin.x
+
+    def min_y(self):
+        return self.normalized().origin.y
+
+    def mid_x(self):
+        return self.origin.x + (self.width / 2)
+
+    def mid_y(self):
+        return self.origin.y + (self.height / 2)
+
+    def max_x(self):
+        norm = self.normalized()
+        return norm.origin.x + norm.width
+
+    def max_y(self):
+        norm = self.normalized()
+        return norm.origin.y + norm.height
+
+    def contract(self, distance):
+        self.origin.x += distance
+        self.origin.y += distance
+        self.width -= 2 * distance
+        self.height -= 2 * distance
+
+
+class SVGRenderer:
```

Comment:
Is there any specific advantage to having free functions? I find having a class for the SVG Renderer to be a more structured way to do it. Also, just in case we need more than one instance of the renderer in the future this could be helpful.

Reply:
in python, if you're just after organization you can use a module (another file).  In this case, however, since this is a plugin to another system, being minimal with some free functions will be best (we don't need to integrate a fully featured library).  Keeping the functionality focused on the need of the integration will help future developers (and/or future you!) understand, debug, and extend the code.

As far as the advantage of free functions, if you write pure free functions (no side effects), then it is clear what each function consumes and returns.  This helps with maintenance since its easier to write unit tests that pass state into a free function than figuring out an additional class that needs to be configured and built in order to use methods.  Its also generally easier to debug, because all the pieces that the function touches are there in the function, it doesn't require also tracking state in another container.

There are other advantages that won't come in to play in this context.

In which case would we need more than one renderer in this plugin?  For rendering multiple svg documents at once?

Reply:
I understand. I'll make the changes accordingly.

> In which case would we need more than one renderer in this plugin? For rendering multiple svg documents at once?

Yes, this is what I thought initially. But now that I think about it we won't generally need to plot more than one timelines with a single function.